### PR TITLE
Proposed GitOps Definition

### PR DESCRIPTION
**Pull request by bricef**
Define proposed GitOps Definition

### CODE REVIEW

File: PRINCIPLES.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+The software systems that we manage vary widely; from battery powered embedded systems driven by microcontrollers, to globally distributed applications with millions of users. 
+Most systems are composed of subsystems themselves made of software.
+It is impossible to comprehensively outline all the specific practices for managing such a variety of systems.
+
+However, despite the myriad differences, several important principles emerge that simplify the task of reliably managing and operating _all_ software systems at scale. GitOps attempts to capture some of these principles in a coherent framework.
+The GitOps principles can be applied to managing entire software systems or applied only to parts of larger systems. 
+
+## Scope
+
+GitOps concerns the verifiable behaviour of computer systems and their interfaces.
```
Comment (tabern) [inline PRINCIPLES.md]: this reads like a measurement - but we mean to say a declaration

File: PRINCIPLES.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+## Scope
+
+GitOps concerns the verifiable behaviour of computer systems and their interfaces. 
+Specifically, GitOps is not about human processes, and is not intended as a model for judging human organisational designs and operational practices.
+
+The GitOps principles are to be used as guiding principles in the development of modern software and system operations. They do not form a concrete specification. 
+
+The GitOps principles are a direction, not a destination. They should be applied pragmatically. For example, whilst they can be applied strictly to an entire software systems, they can also be applied loosely to selected parts of larger systems as part of a progressive adoption.
+
+## The GitOps Principles
```
Comment (tabern) [inline PRINCIPLES.md:81]: is GitOps the correct term still here? Are we tying ourselves too much to Git (software tool)? Flip side - its common and recognizable.

File: PRINCIPLES.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+### 1. Declarative configuration
+
+<div style="border: 1px solid black; padding:5px; font-weight:bold">
+A system managed by GitOps must have its <em>Desired State</em> expressed declaritively as data. 
+ This declarative data must be in a format writable and readable by both humans and software.
```
Comment (tabern) [inline PRINCIPLES.md]: why is human readability required?

File: PRINCIPLES.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+## Introduction
+
+The software systems that we manage vary widely; from battery powered embedded systems driven by microcontrollers, to globally distributed applications with millions of users. 
+Most systems are composed of subsystems themselves made of software.
+It is impossible to comprehensively outline all the specific practices for managing such a variety of systems.
+
+However, despite the myriad differences, several important principles emerge that simplify the task of reliably managing and operating _all_ software systems at scale. GitOps attempts to capture some of these principles in a coherent framework.
+The GitOps principles can be applied to managing entire software systems or applied only to parts of larger systems. 
+
+## Scope
```
Comment (tabern) [inline PRINCIPLES.md:72]: what are the core issues GitOps solves? should we address this?

File: PRINCIPLES.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+### 1. Declarative configuration
+
+<div style="border: 1px solid black; padding:5px; font-weight:bold">
+A system managed by GitOps must have its <em>Desired State</em> expressed declaritively as data. 
+ This declarative data must be in a format writable and readable by both humans and software.
+</div>
+
+<!--
+We've gone back and forth to try to find a really good single-word term for what we mean by "immutable, versioned, declarations of the desired system state".
```
Comment (tabern) [inline PRINCIPLES.md]: i think we need to pick this apart a bit to discuss the declaration of state vs the fulfillment of that declaration. Can we clarify that GitOps is primarily a fulfillment of state declaration of the system? that is, its a mechanism and has nothing to do with the declaration itself?

Comment (tabern) [reply PRINCIPLES.md]: I think this is why 'configuration' falls short - it reads as the declaration rather than the fulfillment of such

File: PRINCIPLES.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+- state store should
+- Business rules and security rules about the acceptable state of the system shoudl be checked against any new proposed version of the desired configuration before such changes are accepted as a new version. (Gates before version)
+- where human process is integrated
+- changes as transactions
+- Even though we say git, it's actually only true when configured in a very particular way.
+- For reliable software operations, It's insufficient to have code define the infrastructure, it must instead be data.
+
+
+
+### 3. Continuous state reconciliation
```
Comment (tabern) [inline PRINCIPLES.md:189]: 3 and 4 seem most critical

File: PRINCIPLES.md [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+- This automation should serve it's human users.
+  To this point, reconciliation automation for GitOps should include a way for humans to temporarily take back the reins as needed.
+
+### 4. Operations through declaration
+
+> When wishing to operate on a software system, a human or software agent will not interact with the running system and modify it directly.
+> Instead, the agent will create a new declarative version of the desired state in the state store.
+
+- All normal operations should occur via the creation of a new uniquely named version, not through direct interaction ith the system under management
+- Break glass - exceptions are acceptable and quite likely. and credentials to access and mutate the system must still be available in almost all cases.
```
Comment (tabern) [inline PRINCIPLES.md:213]: but y tho? Disagree this is acceptable and likely

File: PRINCIPLES.md [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+- Tool and system agnosticicsm.
+  
+- It's ok to be pragmatic with implementations - We recognise that few real systems can be “100% GitOps”.
+  Real world systems have a lifetime measured in years and must interact with many other systems created under alternative paradigms.
+  As such they involve compromises and workarounds.
+  
+- lowest turtle isn't declarative, because the real world isn't.
+  We aim for the lowest turtle to be as small as possible (imperative surface area )
+
+- Git as a satte store: - the "state store" can be a subfolder in a particular branch, and repo.
```
Comment (tabern) [inline PRINCIPLES.md:226]: i feel like this is an implementation detail

File: RATIONALE.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
+# Rationale for GitOps
```
Comment (tabern) [inline RATIONALE.md:1]: this is the problem statement I was missing - need to bring a bit of this up

File: PRINCIPLES.md (hunk 7 above)
Comment (dj1k) [inline PRINCIPLES.md:213]: Let's rephrase this as not something that is acceptable and likely, but instead the goal is to move towards an elimination of all events that trigger manual interventions.

File: PRINCIPLES.md (hunk 2 above)
Comment (tabern) [reply PRINCIPLES.md:81]: New principal: "No out of process access"

consider adding "how to implement these principals"
* all non-gitops changes are tracked and you should drive this down to 0

File: PRINCIPLES.md (hunk 3 above)
Comment (tabern) [reply PRINCIPLES.md]: how do we audit and view the state of the system? We need a principal around this

Comment (tabern) [reply PRINCIPLES.md]: also add some stuff about the ability to enable collaboration between authors

File: PRINCIPLES.md (hunk 5 above)
Comment (tabern) [reply PRINCIPLES.md]: gitops is about the auditability of a declared state, how do you come to consensus across one or more actors that the declared state is correct, and how do you fulfill the declaration of that state.

If we become too dogmatic about the format of that declaration than we risk underminig the applicability of GitOps.

Systems should be able to be closed loop, or human in the loop, to the applciation and preference of the operator.

File: PRINCIPLES.md (hunk 3 above)
Comment (bricef) [reply PRINCIPLES.md]: >why is human readability required?

Operationally, it has been proven time and time  again that system configuration should be human-readable. That's because it encodes the _intent_ of humans. Opaque configuration deprives humans from 1) directly describing their intent and 2) understanding the intent of the configuration. (Insert your own joke here about the opaqueness of YAML).

> how do we audit and view the state of the system?

See details of continuous state reconciliation, which implies observability

> also add some stuff about the ability to enable collaboration between authors

That is a separate and orthogonal concern. We also talk in more detail about collaboration in the section on "Immutable configuration versions", and how the state store is the place where collaboration can be synchronised.

File: PRINCIPLES.md (hunk 5 above)
Comment (bricef) [reply PRINCIPLES.md]: >  GitOps is primarily a fulfillment of state declaration of the system? that is, its a mechanism and has nothing to do with the declaration itself?

Strong disagree. The declaration of the system itself is of key importance and should meet certain criteria to enable the rest of the GitOps loop. The fulfilment of the declaration is the concern of "Continuous state reconciliation" and somewhat "Operations through declaration". 

> gitops is about the auditability of a declared state, 

I agree. I think of this as an observability requirement and tend to say that "the current state of a running system must be observable" to mean the same thing. This is, in my opinion, required for state reconciliation. I didn't add a note there explicitly though, as that section was unfinished.

> how do you come to consensus across one or more actors that the declared state is correct, 

Unless I misunderstand, that's explicitly out-of-scope as it deals with human agents coming to an agreement, which is essentially a human process. I personally have some strong opinions here, but I don't think they belong in a definition of GitOps.
